Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm

```diff
@@ -362,17 +362,22 @@ void didExceedMemoryFootprintThreshold(size_t footprint, const String& domain, u
 
 #if PLATFORM(IOS)
 
-@interface _WKWebsiteDataStoreBSActionHandler : NSObject <_UIApplicationBSActionHandler>
+@interface _WKWebsiteDataStoreBSActionHandler : NSObject <_UIApplicationBSActionHandler> {
+    BlockPtr<WKWebsiteDataStore *(_WKWebPushAction *)> _webPushActionHandler;
+}
 + (_WKWebsiteDataStoreBSActionHandler *)shared;
 - (void)setWebPushActionHandler:(WKWebsiteDataStore *(^)(_WKWebPushAction *action))handler;
+- (BOOL)handleNotificationResponse:(UNNotificationResponse *)response;
 @end
 
 @interface _WKWebsiteDataStoreNotificationCenterDelegate : NSObject <UNUserNotificationCenterDelegate>
 @end
 
 @implementation _WKWebsiteDataStoreNotificationCenterDelegate
+
 - (void)userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)())completionHandler
 {
+    [WKWebsiteDataStore handleNotificationResponse:response];
```

Comment:
Something I remembered is that we have to have this (as well as the BSAction handler that fires the chain of push events) take a UIApp background task for the duration of the event firing. Or we have to provide a completion handler to the client so they can do the equivalent thing. Otherwise the app can get suspended in the middle of event processing.

Reply:
I'm filing a followup radar for this